src/groups/bmq/bmqu/bmqu_managedcallback.h

```diff
@@ -0,0 +1,237 @@
+    /// Construct a callback object of the specified CALLBACK_TYPE in this
+    /// objects' reusable buffer, with the specified `args` passed to its
+    /// constructor.  The buffer must be empty before construction, it is
+    /// the user's responsibility to call `reset()` to destroy the callback.
+    template <class CALLBACK_TYPE, class... ARGS>
+    void createInplace(ARGS&&... args);
+#endif
+
+    /// Store the specified `callback` in this object.
+    /// Note: these setters are slow because they performs function copy, going
+    ///       against the basic idea of ManagedCallback.  They exist solely for
+    ///       the compatibility with other code where this copy is acceptable.
+    ///       In performance-critical paths, `place` should always be used
+    ///       together with reusable ManagedCallback object(s).
+    void set(const VoidFunctor& callback);
+    void set(bslmf::MovableRef<VoidFunctor> callback);
+
+    // ACCESSORS
+    /// Is this object empty or not.
+    bool empty() const;
+
+    /// Call the stored callback.
+    /// The object must be `!empty()` before calling `operator()`.
+    void operator()() const;
+
+  private:
+    // PRIVATE MANIPULATORS
+
+    /// Book and return the allocated memory to store the specified
+    /// `CALLBACK_TYPE` that has to be inherited from `CallbackFunctor`.
+    /// Note that it's the user's responsibility to construct a functor object
+    /// at the provided address.
+    /// The object must be `empty()` before calling `place()`.
+    template <class CALLBACK_TYPE>
+    char* place();
+};
+
+// ============================================================================
+//                             INLINE DEFINITIONS
+// ============================================================================
+
+// ---------------
+// ManagedCallback
+// ---------------
+
+inline ManagedCallback::ManagedCallback(bslma::Allocator* allocator)
+: d_callbackBuffer(allocator)
+, d_empty(true)
+{
+    // NOTHING
+}
+
+inline ManagedCallback::~ManagedCallback()
+{
+    reset();
+}
+
+inline void ManagedCallback::reset()
+{
+    if (!d_empty) {
+        // Not necessary to resize the vector or memset its elements to 0,
+        // we just call the virtual destructor, and `d_empty` flag
+        // prevents us from calling outdated callback.
+        reinterpret_cast<CallbackFunctor*>(d_callbackBuffer.data())
+            ->~CallbackFunctor();
+        d_empty = true;
+    }
+}
+
+template <class CALLBACK_TYPE>
+inline char* ManagedCallback::place()
+{
+    // PRECONDITIONS
+    BSLS_ASSERT_SAFE(d_empty);
+    /// The compilation will fail here on the outer `static_cast` if we
+    /// don't provide a type that is inherited from the base
+    /// `CallbackFunctor` type.
+    /// TODO: replace by static_assert on C++ standard update
+    BSLS_ASSERT_SAFE(0 == static_cast<CALLBACK_TYPE*>(
+                              reinterpret_cast<CallbackFunctor*>(0)));
```

Comment:
Compile time error if we try to pass a template class that is not a child class for `CallbackFunctor`:
```c++
/blazingmq/src/groups/mqb/mqbi/mqbi_dispatcher.h:537:31: error: static_cast from 'CallbackFunctor *' to '(anonymous namespace)::Dummy *', which are not related by inheritance, is not allowed
  537 |         BSLS_ASSERT_SAFE(0 == static_cast<CALLBACK_TYPE*>(
      |                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
  538 |                                   reinterpret_cast<CallbackFunctor*>(0)));
      |                                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```